lms/services/grouping.py

```diff
@@ -0,0 +1,68 @@
+from lms.models import CanvasSection, Grouping
+from lms.models._hashed_id import hashed_id
+
+
+class GroupingService:
+    def __init__(self, db, application_instance_service, course_service):
+        self._db = db
+        self._application_instance = application_instance_service.get()
+        self._course_service = course_service
+
+    def upsert(self, grouping):
+        db_grouping = (
+            self._db.query(Grouping)
+            .filter_by(
+                application_instance=grouping.application_instance,
+                authority_provided_id=grouping.authority_provided_id,
+                type=grouping.type,
+            )
+            .one_or_none()
+        )
+        if not db_grouping:
+            self._db.add(grouping)
+        else:
+            # Update any fields that might have changed
+            db_grouping.lms_name = grouping.lms_name
+            db_grouping.extra = grouping.extra
+
+        return db_grouping or grouping
```

Comment:
So `GroupingService.upsert(grouping)` will either find a matching `Grouping` in the DB and update its `lms_name` and `extra` or, if no matching `Grouping` exists it'll add the giving `grouping` to the DB table.

Seems sensible and I think I like the idea of using the `models.Grouping` class itself as the argument to `upsert()`. It avoids the method having too many arguments like `grouping_svc.upsert(authority_provided_id, type, lms_name, extra)` when those arguments would just be exactly the same as the attributes of `models.Gruoping`. And we can have `upsert_*()` methods for upserting each different type of grouping (Canvas section etc) that call this base `upsert()` method. Those methods don't need to worry about adding the grouping to the DB session or looking for an existing grouping already in the DB. Seems pretty nice :+1: